### ISSUE TITLE
OBJ loading errors

### ISSUE THREAD

**Issue by peardox** [castle-engine/castle-engine#591]
Kenney's models are fairly popular and he produces (some) only as OBJ (but this is far less the case these days)

The following seven packs they offer all fail to load with CGE while a cursory Blender import here and there works fine

Castle Kit
Conveyor Kit
Graveyard Kit
Hexagon Kit
Mini Arena
Minigolf Kit
Modular Buildings

For completeness there's also Food Kit - glass.obj (the only bad one in that pack)

Of particular notice is that some of the models actally are bad in a different way such as Mini Arena - character-soldier.obj. This one imports but has a zero-sized bounding box - Blender just doesn't load that one

As Mini Arena has both types of error it seems a good place to start - https://kenney.nl/assets/mini-arena

In that Kit banner.obj is a sample of other error type (can't read all the data - this is the most common issue)

Across all Kits 542/3353 (16%) OBJs fail (GLTF appears OK though) but that's not the issue - the fact that there's an OBJ load problem is

FYI the GLTF count is 3351 (i.e. 2 total diff but unsure if the overlap is the same) and those all load (Checked with Kenney All In One 1.2.1)

**Comment by michaliskambi** [comment]
( To anyone reading this: To be clear, while we want to support OBJ as good as we can, we advise using glTF file format. If you want to use Kenney's assets with CGE, use the glTF which Kenney provides as well, and ignore OBJ format -- glTF is really better in every way (will load faster, has more features, has clear spec...) See https://castle-engine.io/creating_data_model_formats.php . )

Thanks!

I fixed now `banner.obj` from https://kenney.nl/assets/mini-arena (and anything that failed for the same reason -- the "v" line in OBJ can have > 3 components).

![Zrzut ekranu z 2024-04-01 00-06-11](https://github.com/castle-engine/castle-engine/assets/1076978/6905fffd-dcf4-43f9-8b46-43f52bb2202b)

I tested `glass.obj` from https://kenney.nl/assets/food-kit -- seems to render OK and looks similar to what Blender import of this OBJ looks like. (Which is nice considering the OBJ contains funny lines `vt NaN 0` :) ). Let me know if this fails for you, please show the exact error message and/or screenshot from view3dscene.

![Zrzut ekranu z 2024-04-01 00-16-32](https://github.com/castle-engine/castle-engine/assets/1076978/4a90f880-9568-4a5a-a334-5b49ae9ac55c)

I tested `character-soldier.obj` from https://kenney.nl/assets/mini-arena , it is indeed literally empty in both view3dscene and when importing in Blender, and this is correct, the OBJ file literally specifies nothing:

```
# Created by Kenney (www.kenney.nl)

mtllib character-soldier.mtl
```

I checked a random sample of other models, seem OK. Please let me know if you have other samples that fail.

**Comment by peardox** [comment]
OIK, I'll run all Kenney again 

FYI - https://github.com/peardox/SpritelyFMX (commenting TBD - need to read PasDoc docs...)

**Comment by peardox** [comment]
Close but....

Food Kit\Models\OBJ format\glass.obj (I need to investigate this one but I remember it had a NAN error in Delphi)

Likely all zero (about to check)...
Mini Arena\Models\OBJ format\character-soldier.obj
Mini Dungeon\Models\OBJ format\character-human.obj
Mini Dungeon\Models\OBJ format\character-orc.obj
Mini Skate\Models\OBJ format\character-skate-boy.obj
Mini Skate\Models\OBJ format\character-skate-girl.obj

Note the didn't get rep[orted last time

Anyway ther OBJs are bad so I need to report to Kenney after checking the characters are all zero-sized

**Comment by peardox** [comment]
Hehehe

character-* are all just the .mtl def without any vertex data

Suppose I'd better check DAE as well. Possibly STL? AFAIK that's used a lot in 3D preinting...

**Comment by peardox** [comment]
DAE Results 2526 models - less than OBJ/GLTF

Load Failure for C:\work\Assets\3D\Kenney Game Assets All-in-1 2.1.0\3D assets\Food Kit\Models\DAE format\glass.dae
320 x Warning: Collada: <diffuse> texture refers to missing sampler2D name "_NAME_". Found <image> with the same id, will use it
3000 x Warning: Collada: Missing "S" or "U" parameter (1st component of 2D tex coord) in this <source>

Apart from glass screwing up again 2525 models loaded but with the above warnings and randomly sampled all looked OK

STL Results 2526 models so likely same ones as DAE

All load, look like crap as no colour

**Comment by michaliskambi** [comment]
> Just to stress Food Kit\Models\OBJ format\glass.obj still fails with vt NaN 0
> i.e. view3dscene fixes the issue but CGE doesn't

I added proper handling of `NaN` in tex coords in OBJ -- they will result in nice warning and be reliably replaced with 0. This should be reliable in both FPC and Delphi.

( The OBJ models are still incorrect, whatever generated it -- should not have put `NaN` in OBJ files :) )

Note: This report is getting a bit hard to follow :) If you find something that you believe is CGE bug, please submit a new issue, and make one issue for each separate problem. This way I will not miss anything that should be fixed.

**Comment by peardox** [comment]
I just rebuilt CGE (with issues) and glass stil didn't load - I'll have to check again in a simple test as my code now traps errors and just ignores any bad files - anyway it's only one model so Kenney really needs to fix it.... Same for the character models that are empty

I wasn't sure if the DAE warnings were actually errors or not so just tagged them onto this one

**Comment by michaliskambi** [comment]
> I just rebuilt CGE (with issues) and glass stil didn't load - I'll have to check again in a simple test as my code now traps errors and just ignores any bad files - anyway it's only one model so Kenney really needs to fix it.... Same for the character models that are empty

I tested latest CGE with Delphi 12.0 and it loads glass.obj from https://www.kenney.nl/assets/food-kit OK. 

Here's a screenshot, done using unmodified `examples\viewport_and_scenes\view_3d_model_advanced\`, compiled with Delphi. 

![obraz](https://github.com/castle-engine/castle-engine/assets/1076978/c37104a4-a6ff-4b53-a6e2-ca47e8be6a46)

And the log (`C:\Users\michalis\AppData\Local\view_3d_model_advanced\view_3d_model_advanced.log`) also correctly contains the warning `Warning: Wavefront OBJ: Ignoring invalid "NaN" values in OBJ file`.

If it still fails for you, please describe exactly how -- what do you do (and can you reproduce it with unmodified CGE sample like `view_3d_model_advanced`), and if it fails show exactly how -- what error message and where do you see.

**Comment by peardox** [comment]
Under Delphi I'm getting an error Raised at

https://github.com/castle-engine/castle-engine/blob/f8d9aaa951aec07d40aa403d685cc38d6c7fe849/src/base/castleutils_miscella.inc#L541

And no warning in ther Log

**Comment by peardox** [comment]
And trying view_3d_model_advanced with Delpi 11.3 I get exactly the same issue with glass.obj

**Comment by michaliskambi** [comment]
> Under Delphi I'm getting an error Raised at

That is weird, as it works for me. Are you *sure* you test with latest CGE? The `src/scene/load/x3dloadinternalobj.pas` unit should have a code doing `if SameText(S, 'nan') then`, see here: https://github.com/castle-engine/castle-engine/blob/c700c40cf5d0a3f13ca38a78433c81ee9a6839aa/src/scene/load/x3dloadinternalobj.pas#L451

I added this ~4 days ago, when I mentioned in this thread

> I added proper handling of NaN in tex coords in OBJ -- they will result in nice warning and be reliably replaced with 0. This should be reliable in both FPC and Delphi....

**Comment by peardox** [comment]
Hmm, I'm on 12.1 now and advanced example compiles BUT using TCastleScene.Load(AFile) still gets the NaN EConvert exception

I'll have to write a quick little test

**Comment by peardox** [comment]
Righto - one simple testcase - https://github.com/peardox/Test003

In CastleApp.pas comment out/in lines 73/74 to switch between which model is loaded

Currently it'll try loading the bad obj, switch to the glb and it's fine with a normal model

```
//  AddModel('castle-data:/up.glb');
  AddModel('castle-data:/glass.obj');

```

**Comment by michaliskambi** [comment]
Can you confirm you definitely test with latest CGE, from GitHub master? With the `StrToFloatDotSafeNan` present inside `src/scene/load/x3dloadinternalobj.pas` , looking like this https://github.com/castle-engine/castle-engine/blob/c700c40cf5d0a3f13ca38a78433c81ee9a6839aa/src/scene/load/x3dloadinternalobj.pas#L451 ?

**Comment by peardox** [comment]
Ahahhh - just checked Library paths manually - The IDE Plugin has set the paths for CGE to the old version

I Should be getting Documents\Embarcadero\Studio\23.0\CatalogRepository\castle-engine...
But the Plugin has set everything to Documents\Embarcadero\Studio\22.0\CatalogRepository\castle-engine...

Note the 22 where 23 should be - this is using the 'recommended' install method + Configure Delphi to use Engine

Where's this Path come from? It doesn't appear to be version aware - I've got D11+12 (22.0 + 23.0) with their own CGE's where GetIt would stick them

**Comment by peardox** [comment]
Just pulled latest CGE in the D11 (22.0) dir and now 12 (23.0) works :)

**Comment by peardox** [comment]
Just checking further aand I definitely installed from the right directory

![image](https://github.com/castle-engine/castle-engine/assets/1549219/42481dd0-1999-4cec-8848-fc39c7b55156)

**Comment by peardox** [comment]
Well, that was nasty - just removed D11 and reinstalled CGE but it still wants to use hte now non-existant 22.0 so had to hack the registry to fix a bazillion paths

**Comment by peardox** [comment]
OK - worked out what went wrong
Didn't do the Set CGE path bit (2.3)
Importantly this step is going to fool others as well - could be automated as part of GetIt install?

**Comment by michaliskambi** [comment]
> Where's this Path come from? It doesn't appear to be version aware - I've got D11+12 (22.0 + 23.0) with their own CGE's where GetIt would stick them

The CGE IDE plugin uses CGE path that you explicitly set using menu item "Castle Game Engine -> Set Engine Path".

There's no Delphi-version dependent behavior there, and if people install engine now, by default CGE downloads with default paths -- the engine is not even placed underneath Delphi version-specific subdirectory (and there's no reason for it, we setup path to sources, they will work regardless of Delphi version).

The path doesn't depend on the package location -- since it cannot, I mean when you execute "Castle Game Engine -> ..." menu item operations, you can have open an arbitrary package or project, not necessarily the `AllProjects` that contains the CGE.

Indeed, this would be a trap for GetIt users, as the "Castle Game Engine -> Set Engine Path" is shared between Delphi versions, but installation done by GetIt will not be. Added this to TODO, to address for GetIt distributing.
